Pipes - Angela- Hotel

awilson2017 opened the pull request:
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
| Question | Answer    |
| :------------- | :------------- |
| Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? |   There were many design decision. The first was choosing how many classes I wanted within the module. I chose 3 with Admin class storing the daily functions of a hotel. The next major decision was accepting blocks. I decided to have Blocks Class dual as the reservation function. This was a mistake as I ran into many stumbling blocks and was not able to course correct before the deadline.      |
| Describe a concept that you gained more clarity on as you worked on this assignment. |  I gained clarity on how classes interact with each other. Plus overall design. My ability to take in a large amount of requirements at once and then parse out small manageable tasks will get better in time.   <br><br>Utilizing Git is something I really internalized. For this project I underutilized it and found I wanted to look at code I wrote that was no longer on my computer. |
| Describe a _nominal test_ that you wrote for this assignment. | Creating a list of rooms. I tested for rooms 1 and 20. |
| Describe an _edge case test_ that you wrote for this assignment. |  Creating a list of rooms. I tested for rooms 0 and 21.  |
| How do you feel you did in writing pseudocode first, then writing the tests and then the code? |   I felt okay writing psuedocode. As for writing test and then code, I found with challenging portions of the code, I went straight for coding and then writing the test. Rushing to get the logic out of my mind and onto the screen was important, otherwise I would have forgotten quickly. <br><br>Overall, I consolidated my learning with this assignment and felt like the program could be improve. My assignment is incomplete and the portion I did complete is squirrelly and so I want to sit with someone to go over and specifics to this assignment. It would also be nice to discuss overarching strategies to improve. |

awilson2017 committed:
Create all files for both folders lib and specs

awilson2017 committed:
Write implimentation and specs for Room class. Write test for parameters passed to Room instances.

awilson2017 committed:
Remove Room class and specs because I realized it had no functionalitiy.
Write initialize test for DateRange class and write DateRange implimintation code. Rename Hotel class to Admin class.
Stored unused work in unused-work folder for potential future use.

awilson2017 committed:
Reservation class: add methods reservation_id and total_cost.
Made tests for reservation class' added methods.
Starting to work on Admin class and tests.

awilson2017 committed:
Admin Class: Add methods list_of_rooms, add_reservation, list_reservation(date), and list_vacancies.
Need to write tests for almost all the listed methods.

awilson2017 committed:
realize previous git commits were committed for lib folder.
So specs file was not staged. Specs are written for code except list_of_rooms, add_reservation, list_reservation(date), and list_vacancies methods.

awilson2017 committed:
Write tests for methods list_of_rooms, add_reservation, and list_reservations.

awilson2017 committed:
Write reserve available room by date range method and tests.

awilson2017 committed:
Big Picture: Refactoring Reservation class to dual as Block and Reservation functions.

This commit, refactor the initialize method to take a discount and refactor total_cost method to take discount.

awilson2017 committed:
Refactor Admin class to create method create_block_by_date.

awilson2017 committed:
Admin Class: Add methods create_block_by_date, find_block, list_available_blocked_rooms, and find_rooms_from_block.

Also added method add_reservation_to_block.

awilson2017 committed:
The tests for refactored Admin class were turned off during wave 3's block refactor.
Work on refactoring the specs for block code.

droberts-sea commented on lib/Block.rb:
```diff
@@ -0,0 +1,38 @@
+      @discount_percent = discount_percent
+
+
+      @reservations_array2 = []
```
Why is this called `reservations_array2`? In general, if you need to put a number after a variable's name that's usually a bad sign. Moreover, you don't have a `reservations_array1` that would conflict with it.

droberts-sea replied:
Note that instance variables in different classes (here vs `Hotel`) can have the same name - that's one of the big benefits of classes, is that they keep such things separated.

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+
+      @blocks.each do |block|
+        block.reservations_array2.each do |reservation|
+          rez_by_date << reservation
```
Won't this add all reservations from all blocks to the list, regardless of whether the date is included?

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+      available_rooms = list_of_rooms
+
+      date_range = DateRange.new(check_in, check_out).date_range_array
+
```
All the places that you create a `DateRange`, it seems like you throw away the actual `DateRange` instance and only keep track of the result of calling `date_range_array`. This indicates that a method might be a better choice here, one that takes a checkin and checkout date and returns an array of dates.

The other option would be to write an instance method, something like `DateRange#overlap?(other)`, that takes another `DateRange` and checks to see if they overlap. Then this complex bit of logic would be hidden away in `DateRange` instead of sitting here in `Hotel`.

As long as you've identified "manage a range of dates" as a responsibility, then it probably makes sense to do the later, and keep that responsibility out of the `Hotel` class.

This would also make it much easier to test!

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+
+      num_rooms_to_reserve.length.times do |room_num|
+        block.reservations_array << Hotel::Reservation.new(room_num, check_in, check_out, discount_percent: 0)
+      end
```
This is an example of tight coupling between `Hotel` and `Block`. This method relies on the block having a `reservations_array`, which means if that implementation ever changes this code will need to change too. Instead, it might make sense to write a `reserve_room` method on the `Block` class and call that here.

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+    def create_block_by_date(rooms_per_block, check_in, check_out, block_id, discount_percent: 0.0)
+      room_num_array = self.list_vacancies(check_in, check_out).take(rooms_per_block).to_a
+      @blocks << Hotel::Block.new(room_num_array, check_in, check_out, block_id, discount_percent: 0.0)
+    end
```
I'm not sure this will pass the discount_percent through correctly - did you test this?

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+    # Wave 2 requirement. Add reservation only no block
+    # def add_reservation(room_selection, check_in, check_out)
+    #   @reservations << Hotel::Reservation.new(room_selection, check_in, check_out)
+    # end
```
I like the idea of requiring that all reservations be attached to a block - it simplifies the design somewhat, since you don't need to keep track of reservations independently.

However, it's important that, even if you come up with an interesting design idea, your interface still match the project requirements. Given an instance of `Admin` there still needs to be a method to create a reservation directly. Probably the way to go would be to re-write `add_reservation` to automatically create a block with one room for the given date and then reserve it.

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+      @room_nums = []
+      20.times do |i|
+        @room_nums << (i + 1)
+      end
```
I like that rooms are just integers here. Makes the design much simpler.

droberts-sea commented on lib/DateRange.rb:
```diff
@@ -0,0 +1,27 @@
+
+    def initialize(check_in, check_out)
+      raise ArgumentError.new("Please enter a Check Out date that comes after Check In date.") if check_in > check_out
+
```
Is it valid if `check_in == check_out`?

droberts-sea commented on specs/admin_spec.rb:
```diff
@@ -0,0 +1,274 @@
+    end
+
+    describe "List_of_rooms method" do
+      it "returns of all rooms" do
```
I'm not sure this `describe` block should be nested like this.

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+    def list_available_blocked_rooms(id)
+      block_array = self.find_block(id)
+      block_array.each do |block|
+        return block.room_num_array
```
I'm not convinced this method is necessary. Why couldn't the caller just call `find_block`, and then access the room list from the block?

droberts-sea commented on specs/admin_spec.rb:
```diff
@@ -0,0 +1,274 @@
+        #edge cases
+
+        @admin.list_available_blocked_rooms(2).first.wont_be_kind_of String
+        @admin.list_available_blocked_rooms(2).last.wont_be_kind_of String
```
Edge cases here would be things like:
- What happens when the block has no rooms available?
- What happens if you pass in an ID for a block that doesn't exist?

droberts-sea commented on specs/admin_spec.rb:
```diff
@@ -0,0 +1,274 @@
+    # edge cases
+    @admin.list_vacancies(check_in, check_out).wont_be_instance_of Hotel::Block
+    @admin.list_vacancies(check_in, check_out).wont_be_kind_of Integer
+    @admin.list_vacancies(check_in, check_out).wont_be_kind_of String
```
These are not edge cases - in fact, if the above checks pass, then these will always pass. For example, if `list_vacancies` returned an instance of `Array`, than it can't possibly have returned an instance of `Hotel::Block`.

Instead, edge cases are all about exploring the "edge" between where the input to this method makes sense and where it will cause it to fail. By "input" here I mean both the parameters and the state of the object it's called on. Cases I would want to see tested include:
- What if there are no reservations? (it should return a list of all the rooms)
- Do reservations that don't overlap the date range affect the list? (they shouldn't)
  - Check the boundaries of what counts as an overlap. What about a reservation that ends on the checkin date, or starts on the checkout date?
- What if there are no vacancies? (it should return an empty array)
- What if your date range spans several small reservations? (all of the rooms from those reservations should be omitted from the list)

Each of these might deserve a separate `it` block.

droberts-sea commented on specs/admin_spec.rb:
```diff
@@ -0,0 +1,274 @@
+
+
+        id = 0
+        @admin.find_block(id).wont_be_kind_of Hotel::Block
```
So what will it be? `nil`?

droberts-sea commented on specs/date_range_spec.rb:
```diff
@@ -0,0 +1,56 @@
+
+    it "Returns array of dates in range" do
+      @date_range.date_range_array.must_be_instance_of Array
+    end
```
You should explicitly check that the dates are as expected.

droberts-sea commented on specs/admin_spec.rb:
```diff
@@ -0,0 +1,274 @@
+  before do
+
+    @admin = Hotel::Admin.new
+
```
I like the idea of DRYing all this up with a `before` block, since it's used in most of your tests.

droberts-sea commented on specs/admin_spec.rb:
```diff
@@ -0,0 +1,274 @@
+      @admin.blocks.length.must_equal 1
+
+      # @admin.create_block_by_date(rooms_per_block, check_in, check_out, 2).total_cost.must_equal 200
+      # @admin.create_block_by_date(rooms_per_block, check_in, check_out, 2).wont_equal 0
```
Can you create multiple blocks that share the same dates? What if there are no rooms left for that date?

droberts-sea commented on lib/Admin.rb:
```diff
@@ -0,0 +1,138 @@
+
+    def add_block(room_num_array, check_in, check_out, block_id, discount_percent: 0)
+      @blocks << Hotel::Block.new(room_num_array, check_in, check_out, block_id, discount_percent: 0)
+    end
```
This method is never called, in your product code or in your tests.

droberts-sea commented:
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly	| yes - From your comprehension Qs, it sounds like you weren't happy with your git habits. I like your commit messages, but I think I agree that more frequent commits would be helpful. As far as keeping track of old work attempts, this is a problem solved by branching, a topic you're welcome to go research now but which we'll cover formally in a few weeks.
Answer comprehension questions	| yes
**Design** |
Each class is responsible for a single piece of the program | some - see inline comment re: `DateRange`
Classes are loosely coupled | some - I've identified places inline where tight coupling occurs
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | some - see inline comment
List reservations for a given date | yes
Calculate reservation price | This is commented out? Not sure why.
Invalid date range produces an error | yes
Test coverage | some - reservation tests are missing, tests for `Admin` 
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | no
Test coverage | some - see inline
**Wave 3** |
Create a block of rooms | yes
Check if a block has rooms | I can see how this would be done given existing methods, but there is not an explicit method for this
Reserve a room from a block | yes
Test coverage | some - see inline
**Additional Feedback** |
File Names | The ruby standard is that files should be named using snake_case - all lowercase, underscores between words. So instead of `lib/DateRange.rb` it would be `lib/date_range.rb`.
Test Coverage | In general, your test coverage is not where I'd like it to be. There are many places where you miss interesting test cases and interactions between pieces of your code, which I've tried to call out below. Writing good test cases is not easy, it requires you to have a strong understanding of how your method will flow, how it will be used and what sorts of things might be issues, but that's precisely why we teach it. Please spend some time studying up on testing, and feel free to reach out to us if you want to work one-on-one with someone on this or need some extra problems to work on.
General Vibe | This is a good start, but it feels like there's a way to go still. The general pattern I notice is that your specific code to solve a problem is usually solid, but design and test coverage could be improved. To me, this indicates you're having trouble taking a step back and imagining what your code looks like from the outside. This is an important and difficult skill, and it's one you can grow with deliberate practice.

awilson2017 committed:
try to refactor DateRange class to instructional staff's recommendations. Logic was a bit too much for me to grasp alone. I Will ask for help.
So I decided to finish writing my tests and then will go back and refactor.

awilson2017 committed:
Regarding testing: all tests from the first submission now pass. Working towards refactoring and adding more tests for all classes.

awilson2017 committed:
Add test for calculating both full and discount price

awilson2017 committed:
Admin Class: Refactor create_block_... method to automatically create one reservation.
Refactor add_reservation_to_block method to loosely couple the Admin and Block classes.
Tests write for both refactors.

awilson2017 committed:
complete the desigh_assignemnt.
Some original feedback comments were not addressed for Hotel Revisited. However, the comments I did address deepened my understanding.

awilson2017 committed:
Refactored classes and methods that depended on DateRange instance variable @date_range_array. Now we have loosely coupled classes. DateRange#overlap and DateRange#include are up and running.

awilson2017 committed:
Completely refactored code so functionality is not dependent on DateRange#date_range_array method and completely removed it.
